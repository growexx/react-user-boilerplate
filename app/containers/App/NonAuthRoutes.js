/**
 *
 * App
 *
 * Non auth routes
 */

import React from 'react';
import { Helmet } from 'react-helmet';
import { Switch, Route } from 'react-router-dom';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import AuthContainer from 'containers/Auth/Loadable';
import GlobalStyle from '../../global-styles';

export default function App() {
  return (
    <div>
      <Helmet
        titleTemplate="%s - React.js Boilerplate"
        defaultTitle="React.js Boilerplate"
      >
        <meta name="description" content="A React.js Boilerplate application" />
      </Helmet>
      <Switch>
        <Route path="/login" component={AuthContainer} />
        <Route path="" component={NotFoundPage} />
      </Switch>
      <GlobalStyle />
    </div>
  );
}
