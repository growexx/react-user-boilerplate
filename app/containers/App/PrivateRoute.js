/* eslint-disable space-before-function-paren */
import React from 'react';
import { Route, Redirect } from 'react-router';
import PropTypes from 'prop-types';
import { userExists } from '../../utils/userExists';

const PrivateRoute = mainProps => {
  const { component: Component, ...rest } = mainProps;

  return (
    <Route
      {...rest}
      render={props =>
        userExists() ? (
          <Component {...props} />
        ) : (
          <Redirect
            to={{
              pathname: '/login',
            }}
          />
        )
      }
    />
  );
};

PrivateRoute.propTypes = {
  component: PropTypes.func,
};

export default PrivateRoute;
