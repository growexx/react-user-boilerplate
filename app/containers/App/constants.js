/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

import {
  FAV_ICON_152_152,
  FAV_ICON_16_16,
  FAV_ICON_32_32,
  APPLE_TOUCH_ICON,
  FAV_ICON_192_192,
} from '../../images/favicons';
export const LOAD_REPOS = 'boilerplate/App/LOAD_REPOS';
export const LOAD_REPOS_SUCCESS = 'boilerplate/App/LOAD_REPOS_SUCCESS';
export const LOAD_REPOS_ERROR = 'boilerplate/App/LOAD_REPOS_ERROR';
export const LOAD_APP = 'boilerplate/App/LOAD_APP';
export const FAV_ICONS = [
  {
    rel: 'icon',
    type: 'image/png',
    sizes: '512x512',
    href: FAV_ICON_152_152,
  },

  {
    rel: 'icon',
    type: 'image/png',
    sizes: '192x192',
    href: FAV_ICON_192_192,
  },
  {
    rel: 'icon',
    type: 'image/png',
    sizes: '32x32',
    href: FAV_ICON_32_32,
  },
  {
    rel: 'icon',
    type: 'image/png',
    sizes: '16x16',
    href: FAV_ICON_16_16,
  },
  {
    rel: 'apple-touch-icon',
    href: APPLE_TOUCH_ICON,
  },
];
