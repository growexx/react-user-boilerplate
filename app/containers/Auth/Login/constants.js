export const CHANGE_EMAIL = 'boilerplate/Login/CHANGE_EMAIL';
export const CHANGE_PASSWORD = 'boilerplate/Login/CHANGE_PASSWORD';
export const ERROR = 'boilerplate/Login/ERROR';
export const LOADING = 'boilerplate/Login/LOADING';
export const SUCCESS = 'boilerplate/Login/SUCCESS';
export const LOGIN = 'boilerplate/Login/LOGIN';
export const RESET = 'boilerplate/Login/RESET';
export const GOOGLE_LOGIN = 'boilerplate/Login/GOOGLELOGIN';
export const FACEBOOK_LOGIN = 'boilerplate/Login/FACEBOOKLOGIN';
