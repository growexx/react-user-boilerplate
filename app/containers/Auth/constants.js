export const AUTH_TYPE = ['LOGIN', 'REGISTER'];

// API-ROUTES
export const API_URL = 'https://b0257f3aa104.ngrok.io';
export const AUTH = 'auth';
export const API_ENDPOINTS = {
  LOGIN: `${API_URL}/${AUTH}/signin`,
  LIST: 'https://randomuser.me/api/?results=5&inc=name,gender,email,nat&noinfo',
  LIST_AVATAR:
    'https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png',
  EXPORT_CSV: '',
};
