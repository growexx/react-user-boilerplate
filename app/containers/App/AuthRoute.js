/* eslint-disable space-before-function-paren */
import React from 'react';
import { Route, Redirect } from 'react-router';
import PropTypes from 'prop-types';
import { userExists } from 'utils/Helper';
import { ROUTES } from '../constants';

const AuthRoute = mainProps => {
  const { component: Component, ...rest } = mainProps;

  return (
    <Route
      {...rest}
      render={props =>
        !userExists() ? (
          <Component {...props} />
        ) : (
          <Redirect
            to={{
              pathname: ROUTES.HOME,
            }}
          />
        )
      }
    />
  );
};

AuthRoute.propTypes = {
  component: PropTypes.func,
};

export default AuthRoute;
