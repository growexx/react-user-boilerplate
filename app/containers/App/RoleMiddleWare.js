import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
/* import { Spin } from 'antd'; */
import {
  RESTRICTED_ROUTES,
  ROLE_BASED_ROUTE_ACCESS,
  ROLE_BASED_DEFAULT_ROUTE,
  ROUTES,
} from '../constants';
import { USER_DATA_KEY } from '../../utils/constants';
import StorageService from '../../utils/StorageService';
import { loginSuccessResponse } from '../Auth/Login/stub/login.stub';

class RoleMiddleWare extends React.Component {
  /**
   * NOTE: for actual implementation of the component, remove comments with following note  "ACTUAL API INTEGRATION CODE" and remove demo code
   */
  constructor(props) {
    super(props);
    this.state = {
      loader: true,
      userData: {},
      isRestrictedRoute: true,
    };
  }

  // Get role routes
  doesRoleHaveRouteAccess = (role, route) =>
    (ROLE_BASED_ROUTE_ACCESS[role || 'user'] || []).includes(route);

  static getDerivedStateFromProps(props, state) {
    // Only check if route is restricted or not
    // Only Reset isRestrictedRoute if userData is not fetched
    if (!state.userData || !Object.keys(state.userData).length) {
      const isRestrictedRoute = RESTRICTED_ROUTES.includes(props.path);

      return {
        ...state,
        isRestrictedRoute,
        loader: isRestrictedRoute,
      };
    }
    return state;
  }

  /**
   * Fetch user data and store in local state
   */
  /**
    /* ACTUAL API INTEGRATION CODE
  fetchUserRole = () => {
  
     const data = {
        method: 'GET',
      };
      const requestURL = `${API_ENDPOINTS.USER_DETAILS_API}`;
      request(requestURL, data).then(response => {
        if (response.status) {
          // Save in local storage
          StorageService.set(USER_DATA_KEY, response.data);
          this.setState({ userData: response.data }, () => {
            this.takeDecision((response.data && response.data.role) || '');
          });
        } else {
          logout();
          this.props.history.push('/login');
        }
      });
    }
  };
       */

  /**
   * Takes decision and stop the loader
   */
  takeDecision = role => {
    this.setState({ loader: false });
    /**
     * doesRoleHaveRouteAccess (role, path)
     * Check which path do you need to pass based on paths defined in constant file
     */

    if (!this.doesRoleHaveRouteAccess(role, this.props.path)) {
      if (this.props.showError) {
        this.props.history.push(ROUTES.UNAUTHORIZED);
      } else {
        this.props.history.push(ROLE_BASED_DEFAULT_ROUTE[role]);
      }
    }
  };

  // On component load get user data
  componentDidMount() {
    /**
     * ACTUAL API INTEGRATION CODE
    if (this.state.isRestrictedRoute) {
      this.fetchUserRole();
    }
    */
    // -------------Demo--------------------
    const response = loginSuccessResponse;
    // Save in local storage
    StorageService.set(USER_DATA_KEY, response.data);
    this.setState({ userData: response.data }, () => {
      this.takeDecision((response.data && response.data.role) || '');
    });
    // -------------Demo--------------------
  }

  render() {
    const { component: Component, ...rest } = this.props;
    return (
      <Component {...rest} Component={null} userData={this.state.userData} />
    );
    // Note: Remove in actual Code
    // -------------Demo--------------------

    /**
     * ACTUAL API INTEGRATION CODE
    return this.state.loader ? (
      <Spin spinning={this.state.loader} size="large" />
    ) : (
      <Component {...rest} Component={null} userData={this.state.userData} />
    );
    */
  }
}

RoleMiddleWare.propTypes = {
  history: PropTypes.any,
  childProps: PropTypes.any,
  component: PropTypes.any,
  path: PropTypes.string,
  showError: PropTypes.bool,
};

export default withRouter(RoleMiddleWare);
