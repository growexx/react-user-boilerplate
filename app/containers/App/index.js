/* eslint-disable react/no-array-index-key */
/**
 *
 * App
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 */

import React, { useEffect } from 'react';
import { Helmet } from 'react-helmet';
import styled from 'styled-components';
import { Switch, Route } from 'react-router-dom';
import HomePage from 'containers/HomePage/Loadable';
import Profile from 'containers/Profile/Loadable';
import FeaturePage from 'containers/FeaturePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import UnauthorizedPage from 'containers/UnauthorizedPage/Loadable';
import Login from 'containers/Auth/Login/Loadable';
import TwoFactorAuthentication from 'containers/Auth/TwoFactorAuthentication/Loadable';
import Logout from 'containers/Auth/Logout/Loadable';
import Loader from 'examples/ListLoader/Loadable';
import Register from 'containers/Auth/Registration/Loadable';
import ExportDataToCsv from 'examples/ExportDataToCsv/Loadable';
import Users from 'examples/Users/Loadable';
import Charts from 'examples/Charts/Loadable';
import Products from 'examples/Products/Loadable';
import SampleForm from 'examples/SampleForm/Loadable';
import ChangePassword from 'containers/ChangePassword/Loadable';
import MultiTabSupport from 'examples/MultiTabSupport/Loadable';
import ForgotPassword from 'containers/Auth/ForgotPassword/Loadable';
import NumeralConversion from 'examples/NumeralConversion/Loadable';
import { FAV_ICONS } from './constants';
import PrivateRoute from './PrivateRoute';
import RoleMiddleWare from './RoleMiddleWare';
import AuthRoute from './AuthRoute';
import GlobalStyle from '../../global-styles';
import { ROUTES } from '../constants';
import { manageSession } from '../../utils/Helper';

const AppWrapper = styled.div`
  display: flex;
  min-height: 100vh;
  flex-direction: column;
`;

export default function App() {
  useEffect(() => {
    window.addEventListener('storage', manageSession, []);
    return () => {
      window.removeEventListener('storage', window);
    };
  }, []);

  return (
    <AppWrapper data-testid="AppRoutes">
      <Helmet
        titleTemplate="%s - React.js Boilerplate"
        defaultTitle="React.js Boilerplate"
      >
        <meta name="description" content="A React.js Boilerplate application" />
        {FAV_ICONS.map((favIcon, index) => (
          <link {...favIcon} key={index} />
        ))}
      </Helmet>
      <Switch>
        <PrivateRoute exact path={ROUTES.HOME} component={FeaturePage} />
        <PrivateRoute path={ROUTES.GITHUB_SEARCH} component={HomePage} />
        <PrivateRoute path={ROUTES.PROFILE} component={Profile} />
        <PrivateRoute path={ROUTES.LOGOUT} component={Logout} />
        <PrivateRoute path={ROUTES.LOADER} component={Loader} />
        <PrivateRoute path={ROUTES.EXPORT_DATA} component={ExportDataToCsv} />
        <PrivateRoute path={ROUTES.USERS} component={Users} />
        <PrivateRoute path={ROUTES.CHARTS} component={Charts} />
        <PrivateRoute path={ROUTES.PRODUCTS} component={Products} />
        <PrivateRoute
          path={ROUTES.MULTI_TAB_SUPPORT}
          component={MultiTabSupport}
        />
        <PrivateRoute
          path={ROUTES.CHANGE_PASSWORD}
          component={ChangePassword}
        />
        <RoleMiddleWare
          path={ROUTES.TEST_ADMIN_PAGE}
          component={() => <div>This is Admin Role Page</div>}
          // ShowError redirects to 403
          showError
        />
        <PrivateRoute path={ROUTES.SAMPLE_FORM} component={SampleForm} />
        <PrivateRoute
          path={ROUTES.NUMERAL_CONVERTER}
          component={NumeralConversion}
        />
        <AuthRoute exact path={ROUTES.LOGIN} component={Login} />
        <AuthRoute
          exact
          path={ROUTES.TWO_FACTOR_AUTHENTICATION}
          component={TwoFactorAuthentication}
        />
        <AuthRoute exact path={ROUTES.REGISTER} component={Register} />
        <AuthRoute
          exact
          path={ROUTES.FORGOT_PASSWORD}
          component={ForgotPassword}
        />
        <Route exact path={ROUTES.UNAUTHORIZED} component={UnauthorizedPage} />
        <Route path="" component={NotFoundPage} />
      </Switch>
      <GlobalStyle />
    </AppWrapper>
  );
}
