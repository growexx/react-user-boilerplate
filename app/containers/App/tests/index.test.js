import React from 'react';
import { Provider } from 'react-redux';
import { render } from 'react-testing-library';
import ShallowRenderer from 'react-test-renderer/shallow';
import { browserHistory, MemoryRouter } from 'react-router-dom';
import { IntlProvider } from 'react-intl';
import App from '../index';
import configureStore from '../../../configureStore';
import { ROUTES } from '../../constants';
import StorageService from '../../../utils/StorageService';
import { TOKEN_KEY } from '../../../utils/constants';

let store;
const renderer = new ShallowRenderer();
const componentWrapper = () =>
  render(
    <Provider store={store}>
      <IntlProvider locale="en">
        <MemoryRouter initialEntries={[ROUTES.TEST_ADMIN_PAGE]}>
          <App />
        </MemoryRouter>
      </IntlProvider>
    </Provider>,
  );

describe('<App />', () => {
  beforeAll(() => {
    store = configureStore({}, browserHistory);
    StorageService.set(TOKEN_KEY, 'TOKENVALUE');
  });
  it('should render and match the snapshot', () => {
    renderer.render(<App />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
  it('should render Div', () => {
    const { container } = componentWrapper();
    const element = container.firstElementChild;
    expect(element.tagName).toEqual('DIV');
  });
});
