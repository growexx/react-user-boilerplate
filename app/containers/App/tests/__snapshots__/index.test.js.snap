// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render and match the snapshot 1`] = `
<ForwardRef
  data-testid="AppRoutes"
>
  <HelmetWrapper
    defaultTitle="React.js Boilerplate"
    defer={true}
    encodeSpecialCharacters={true}
    titleTemplate="%s - React.js Boilerplate"
  >
    <meta
      content="A React.js Boilerplate application"
      name="description"
    />
    <link
      href="IMAGE_MOCK"
      rel="icon"
      sizes="512x512"
      type="image/png"
    />
    <link
      href="IMAGE_MOCK"
      rel="icon"
      sizes="192x192"
      type="image/png"
    />
    <link
      href="IMAGE_MOCK"
      rel="icon"
      sizes="32x32"
      type="image/png"
    />
    <link
      href="IMAGE_MOCK"
      rel="icon"
      sizes="16x16"
      type="image/png"
    />
    <link
      href="IMAGE_MOCK"
      rel="apple-touch-icon"
    />
  </HelmetWrapper>
  <Switch>
    <PrivateRoute
      component={[Function]}
      exact={true}
      path="/"
    />
    <PrivateRoute
      component={[Function]}
      path="/github-search"
    />
    <PrivateRoute
      component={[Function]}
      path="/profile"
    />
    <PrivateRoute
      component={[Function]}
      path="/logout"
    />
    <PrivateRoute
      component={[Function]}
      path="/loader"
    />
    <PrivateRoute
      component={[Function]}
      path="/export-data"
    />
    <PrivateRoute
      component={[Function]}
      path="/users"
    />
    <PrivateRoute
      component={[Function]}
      path="/charts"
    />
    <PrivateRoute
      component={[Function]}
      path="/products"
    />
    <PrivateRoute
      component={[Function]}
      path="/multi-tab-support"
    />
    <PrivateRoute
      component={[Function]}
      path="/change-password"
    />
    <withRouter(RoleMiddleWare)
      component={[Function]}
      path="/admin"
      showError={true}
    />
    <PrivateRoute
      component={[Function]}
      path="/sample-form"
    />
    <PrivateRoute
      component={[Function]}
      path="/number-converter-demo"
    />
    <AuthRoute
      component={[Function]}
      exact={true}
      path="/login"
    />
    <AuthRoute
      component={[Function]}
      exact={true}
      path="/two-factor-authentication"
    />
    <AuthRoute
      component={[Function]}
      exact={true}
      path="/register"
    />
    <AuthRoute
      component={[Function]}
      exact={true}
      path="/forgot-password"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/403"
    />
    <Route
      component={[Function]}
      path=""
    />
  </Switch>
  <GlobalStyleComponent />
</ForwardRef>
`;
